code/scripts/SourceCodeReader.hs

```diff
@@ -0,0 +1,86 @@
+module SourceCodeReader (extractEntryData) where
+
+import Data.List
+import System.IO
+import System.Directory
+import qualified Data.Text as T
+
+import DirectoryController as DC (FileName)
+
+type DataName = String
+type NewtypeName = String
+type ClassName = String
+type DtNtName = String
+
+-- extracts data, newtype and class names + instances (new data-oriented format)
+extractEntryData :: DC.FileName -> FilePath -> 
+  IO ([DataName],[NewtypeName],[ClassName],[(DtNtName,ClassName)])
```

Comment:
It would be better to create a 4-member record with fields instead of a 4-tuple. Also, you can then make the fields have a `!` in front of them, making them strict. That ought to solve your `seq` and `forceRead` problem more elegantly.

Reply:
I created a new `EntryData` data type with the strict fields. However, I am still encountering the same delayed read on closed handle error.

Reply:
Well, the record is still a good thing. I'll have to think harder about the strictness problem - so leave in your `seq`s for now.